core/sparse_array.cpp

```diff
@@ -68,6 +88,31 @@ SparseArray::~SparseArray(void)
 	
 	free(strengths_);
 	strengths_ = nullptr;
+
```

Comment:
Proposal: re-use this stuff: don't free/alloc, instead just clear the memory and reuse the data structure.

Reply:
Yes, might be a good idea, if the allocations/frees show up as significant in a profile.  This is the destructor for SparseArray, though, so it pretty much has to deallocate the memory it was using (perhaps it could return it to a shared pool, but that might be more work than it is worth since the blocks are not of a set fixed size).  So if we want to get into this optimization, probably the right level for it to happen at would be reuse of the SparseArray object itself.  That could probably be done inside InteractionType.  Food for thought, depending on whether this shows up as a bottleneck.